File: python/test_webhook.py
"""
Minimal webhook example for Crebain API.

Usage:
    1. Install dependencies: pip install flask crebain-client
    2. Run this server: python test_webhook.py
    3. Expose with ngrok: ngrok http 5000
    4. Register webhook with the ngrok URL

For local testing without a public URL, use the --register flag
with your ngrok URL after starting ngrok.
"""

import os
import json
import logging
from flask import Flask, request, jsonify
from crebain_client import CrebainClient, verify_signature

# Configuration - set these via environment variables
WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET", "")
CREBAIN_API_KEY = os.getenv("CREBAIN_API_KEY", "")
CREBAIN_BASE_URL = os.getenv("CREBAIN_BASE_URL", "")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY", "")

# Logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s | %(message)s")
logger = logging.getLogger(__name__)

app = Flask(__name__)


@app.route("/webhook", methods=["POST"])
def handle_webhook():
    """Handle incoming webhook from Crebain API."""

    # Get signature headers
    timestamp = request.headers.get("X-Crebain-Timestamp")
    signature = request.headers.get("X-Crebain-Signature")
    raw_body = request.get_data()

    logger.info("=" * 50)
    logger.info("Received webhook")
    logger.info("  Timestamp: %s", timestamp)
    logger.info("  Signature: %s", signature[:50] + "..." if signature else None)

    # Verify signature
    if not verify_signature(WEBHOOK_SECRET, timestamp, raw_body, signature):
        logger.error("Invalid signature!")
        return jsonify({"error": "Invalid signature"}), 401

    logger.info("  Signature: VALID")

    # Parse and log the event
    event = request.json
    event_type = event.get("event")
    request_id = event.get("request_id")
    org_id = event.get("org_id")
    kind = event.get("kind")

    logger.info("  Event: %s", event_type)
    logger.info("  Request ID: %s", request_id)
    logger.info("  Org ID: %s", org_id)
    logger.info("  Kind: %s", kind)

    if event_type == "request.complete":
        result = event.get("result", {})
        logger.info("  Result: %s", json.dumps(result, indent=2)[:500])

        # Here you would typically:
        # 1. Update your database
        # 2. Notify your users
        # 3. Trigger downstream processing

    logger.info("=" * 50)
    return jsonify({"status": "ok"}), 200


@app.route("/health", methods=["GET"])
def health():
    """Health check endpoint."""
    return jsonify({"status": "healthy"}), 200


def register_webhook(webhook_url: str):
    """Register a webhook with the Crebain API."""
    client = CrebainClient(api_key=CREBAIN_API_KEY, base_url=CREBAIN_BASE_URL, supabase_anon_key=SUPABASE_ANON_KEY)

    try:
        webhook = client.create_webhook(
            url=webhook_url,
            secret=WEBHOOK_SECRET,
        )
        logger.info("Webhook registered successfully!")
        logger.info("  Webhook ID: %s", webhook.id)
        logger.info("  URL: %s", webhook.url)
        return webhook
    except Exception as e:
        logger.error("Failed to register webhook: %s", e)
        raise


def list_webhooks():
    """List all registered webhooks."""
    client = CrebainClient(api_key=CREBAIN_API_KEY, base_url=CREBAIN_BASE_URL, supabase_anon_key=SUPABASE_ANON_KEY)

    webhooks = client.list_webhooks()
    logger.info("Registered webhooks:")
    for wh in webhooks:
        logger.info("  - %s: %s (enabled: %s)", wh.id, wh.url, wh.enabled)
    return webhooks


if __name__ == "__main__":
    import sys

    if len(sys.argv) > 1:
        command = sys.argv[1]

        if command == "--register" and len(sys.argv) > 2:
            # Register webhook: python test_webhook.py --register https://your-ngrok-url.ngrok.io/webhook
            webhook_url = sys.argv[2]
            register_webhook(webhook_url)

        elif command == "--list":
            # List webhooks: python test_webhook.py --list
            list_webhooks()

        else:
            print("Usage:")
            print("  python test_webhook.py              # Start webhook server")
            print("  python test_webhook.py --register <url>  # Register webhook")
            print("  python test_webhook.py --list       # List webhooks")
    else:
        # Start the Flask server
        port = int(os.getenv("PORT", "5001"))
        logger.info("Starting webhook server on http://localhost:%d", port)
        logger.info("Webhook endpoint: http://localhost:%d/webhook", port)
        logger.info("")
        logger.info("To expose publicly, run: ngrok http %d", port)
        logger.info("Then register: python test_webhook.py --register https://xxx.ngrok.io/webhook")
        app.run(host="0.0.0.0", port=port, debug=True)

File: java/src/main/java/com/crebain/client/example/TestWebhook.java
package com.crebain.client.example;

import com.crebain.client.CrebainClient;
import com.crebain.client.model.WebhookSubscription;
import com.crebain.client.webhook.WebhookVerifier;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.*;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.logging.*;

/**
 * Minimal webhook example for Crebain API (Java).
 * <p>
 * Usage:
 *   1. Build: mvn package
 *   2. Run server: java -cp target/crebain-client-1.0.0.jar com.crebain.client.example.TestWebhook
 *   3. Expose with ngrok: ngrok http 5001
 *   4. Register webhook: java -cp target/crebain-client-1.0.0.jar com.crebain.client.example.TestWebhook --register https://xxx.ngrok.io/webhook
 * <p>
 * Commands:
 *   (no args)           - Start webhook server on port 5001
 *   --register <url>    - Register a webhook with the given URL
 *   --list              - List all registered webhooks
 */
public class TestWebhook {

    private static final Logger logger = Logger.getLogger(TestWebhook.class.getName());
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    // Configuration - set via environment variables
    private static final String WEBHOOK_SECRET = getEnvOrDefault("WEBHOOK_SECRET", "");
    private static final String CREBAIN_API_KEY = getEnvOrDefault("CREBAIN_API_KEY", "");
    private static final String CREBAIN_BASE_URL = getEnvOrDefault("CREBAIN_BASE_URL", "");
    private static final String SUPABASE_ANON_KEY = getEnvOrDefault("SUPABASE_ANON_KEY", "");

    public static void main(String[] args) {
        setupLogging();

        if (args.length > 0) {
            String command = args[0];

            if ("--register".equals(command) && args.length > 1) {
                // Register webhook
                String webhookUrl = args[1];
                registerWebhook(webhookUrl);
            } else if ("--list".equals(command)) {
                // List webhooks
                listWebhooks();
            } else {
                printUsage();
            }
        } else {
            // Start the webhook server
            startServer();
        }
    }

    private static void setupLogging() {
        Logger rootLogger = Logger.getLogger("");
        rootLogger.setLevel(Level.INFO);

        for (Handler handler : rootLogger.getHandlers()) {
            handler.setLevel(Level.INFO);
            handler.setFormatter(new SimpleFormatter() {
                @Override
                public String format(LogRecord record) {
                    return String.format("%1$tF %1$tT | %2$s%n", record.getMillis(), record.getMessage());
                }
            });
        }
    }

    private static String getEnvOrDefault(String name, String defaultValue) {
        String value = System.getenv(name);
        return (value != null && !value.isEmpty()) ? value : defaultValue;
    }

    private static void startServer() {
        int port = Integer.parseInt(getEnvOrDefault("PORT", "5001"));

        try {
            HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
            server.createContext("/webhook", TestWebhook::handleWebhook);
            server.createContext("/health", TestWebhook::handleHealth);
            server.setExecutor(null);

            logger.info("Starting webhook server on http://localhost:" + port);
            logger.info("Webhook endpoint: http://localhost:" + port + "/webhook");
            logger.info("");
            logger.info("To expose publicly, run: ngrok http " + port);
            logger.info("Then register: java -cp target/crebain-client-1.0.0.jar " +
                    "com.crebain.client.example.TestWebhook --register https://xxx.ngrok.io/webhook");

            server.start();
        } catch (IOException e) {
            logger.severe("Failed to start server: " + e.getMessage());
            System.exit(1);
        }
    }

    private static void handleWebhook(HttpExchange exchange) throws IOException {
        if (!"POST".equals(exchange.getRequestMethod())) {
            sendResponse(exchange, 405, "{\"error\": \"Method not allowed\"}");
            return;
        }

        // Get signature headers
        String timestamp = exchange.getRequestHeaders().getFirst("X-Crebain-Timestamp");
        String signature = exchange.getRequestHeaders().getFirst("X-Crebain-Signature");

        // Read raw body
        byte[] rawBody = exchange.getRequestBody().readAllBytes();

        logger.info("==================================================");
        logger.info("Received webhook");
        logger.info("  Timestamp: " + timestamp);
        logger.info("  Signature: " + (signature != null ? signature.substring(0, Math.min(50, signature.length())) + "..." : null));

        // Verify signature
        if (!WebhookVerifier.verify(WEBHOOK_SECRET, timestamp, rawBody, signature)) {
            logger.severe("  Signature: INVALID");
            sendResponse(exchange, 401, "{\"error\": \"Invalid signature\"}");
            return;
        }

        logger.info("  Signature: VALID");

        // Parse and log the event
        String bodyString = new String(rawBody, StandardCharsets.UTF_8);
        JsonObject event = JsonParser.parseString(bodyString).getAsJsonObject();

        String eventType = event.has("event") ? event.get("event").getAsString() : null;
        String requestId = event.has("request_id") ? event.get("request_id").getAsString() : null;
        String orgId = event.has("org_id") ? event.get("org_id").getAsString() : null;
        String kind = event.has("kind") ? event.get("kind").getAsString() : null;

        logger.info("  Event: " + eventType);
        logger.info("  Request ID: " + requestId);
        logger.info("  Org ID: " + orgId);
        logger.info("  Kind: " + kind);

        if ("request.complete".equals(eventType)) {
            if (event.has("result") && !event.get("result").isJsonNull()) {
                String resultStr = gson.toJson(event.get("result"));
                logger.info("  Result: " + resultStr.substring(0, Math.min(500, resultStr.length())));
            }

            // Here you would typically:
            // 1. Update your database
            // 2. Notify your users
            // 3. Trigger downstream processing
        }

        logger.info("==================================================");
        sendResponse(exchange, 200, "{\"status\": \"ok\"}");
    }

    private static void handleHealth(HttpExchange exchange) throws IOException {
        sendResponse(exchange, 200, "{\"status\": \"healthy\"}");
    }

    private static void sendResponse(HttpExchange exchange, int statusCode, String body) throws IOException {
        byte[] responseBytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(statusCode, responseBytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(responseBytes);
        }
    }

    private static void registerWebhook(String webhookUrl) {
        if (CREBAIN_API_KEY.isEmpty() || CREBAIN_BASE_URL.isEmpty()) {
            logger.severe("Missing CREBAIN_API_KEY or CREBAIN_BASE_URL environment variables");
            System.exit(1);
        }

        if (WEBHOOK_SECRET.isEmpty()) {
            logger.severe("Missing WEBHOOK_SECRET environment variable");
            System.exit(1);
        }

        try (CrebainClient client = CrebainClient.builder()
                .apiKey(CREBAIN_API_KEY)
                .baseUrl(CREBAIN_BASE_URL)
                .supabaseAnonKey(SUPABASE_ANON_KEY)
                .timeout(Duration.ofSeconds(30))
                .build()) {

            WebhookSubscription webhook = client.createWebhook(webhookUrl, WEBHOOK_SECRET);

            logger.info("Webhook registered successfully!");
            logger.info("  Webhook ID: " + webhook.getId());
            logger.info("  URL: " + webhook.getUrl());

        } catch (Exception e) {
            logger.severe("Failed to register webhook: " + e.getMessage());
            System.exit(1);
        }
    }

    private static void listWebhooks() {
        if (CREBAIN_API_KEY.isEmpty() || CREBAIN_BASE_URL.isEmpty()) {
            logger.severe("Missing CREBAIN_API_KEY or CREBAIN_BASE_URL environment variables");
            System.exit(1);
        }

        try (CrebainClient client = CrebainClient.builder()
                .apiKey(CREBAIN_API_KEY)
                .baseUrl(CREBAIN_BASE_URL)
                .supabaseAnonKey(SUPABASE_ANON_KEY)
                .timeout(Duration.ofSeconds(30))
                .build()) {

            List<WebhookSubscription> webhooks = client.listWebhooks();

            logger.info("Registered webhooks:");
            for (WebhookSubscription wh : webhooks) {
                logger.info("  - " + wh.getId() + ": " + wh.getUrl() + " (enabled: " + wh.isEnabled() + ")");
            }

        } catch (Exception e) {
            logger.severe("Failed to list webhooks: " + e.getMessage());
            System.exit(1);
        }
    }

    private static void printUsage() {
        System.out.println("Usage:");
        System.out.println("  java -cp target/crebain-client-1.0.0.jar com.crebain.client.example.TestWebhook");
        System.out.println("      Start webhook server");
        System.out.println("");
        System.out.println("  java -cp target/crebain-client-1.0.0.jar com.crebain.client.example.TestWebhook --register <url>");
        System.out.println("      Register webhook");
        System.out.println("");
        System.out.println("  java -cp target/crebain-client-1.0.0.jar com.crebain.client.example.TestWebhook --list");
        System.out.println("      List webhooks");
    }
}
